### ISSUE TITLE
Performance issue 

### ISSUE THREAD

**Issue by JonWallsten** [angular/angular#45055]
### Which @angular/* package(s) are the source of the bug?

animations

### Is this a regression?

No

### Description

When doing some performance debugging in our application I noticed that around 1/3 of the time to process our components is used by a single function related to animations.

https://github.com/angular/angular/blob/4c778cdb28de128c6ddecd2fd6cb6257d675500b/packages/animations/browser/src/render/transition_animation_engine.ts#L855

If you look at my performance profile you can see that the anonymous function marked in red (and the ones next to it) is calling `_balanceNamespaceList` and it takes almost a third of the total processing for rendering the layout.
![image](https://user-images.githubusercontent.com/5993909/153586425-038c048b-d0b8-461e-9899-ea94f32657a9.png)

Is it supposed to be like this? Can it be optimized? Or can be disable something so it doesn't do this without removing all animations?

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 13.2.1
Node: 16.14.0
Package Manager: npm 8.3.1
OS: win32 x64

Angular: 13.2.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, material, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1302.1
@angular-devkit/build-optimizer   0.1302.1
@angular-devkit/core              13.2.1
@angular-devkit/schematics        13.2.1
@ngtools/webpack                  13.2.1
@schematics/angular               13.2.1
ng-packagr                        13.2.1
rxjs                              7.5.2
typescript                        4.5.5
webpack                           5.67.0
```

**Comment by JoostK** [comment]
I can think of an optimization that wouldn't be too hard to implement, but it's hard to see the effect without a good repro.

**Comment by JonWallsten** [comment]
It's going to be hard to make a repro since this is a pretty huge and complex app with complex templates and I'm not allowed to make it public. The expensive part in this code seems to be the containsElement calls. Is there anyway I can test the code for you?

**Comment by JoostK** [comment]
Yeah I can try what I think will work and then share it with you (likely using a PR artefact)

**Comment by JoostK** [comment]
Here you go: #45057. Not sure if the build will pass but once it does you can [follow these steps](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#getting-packages-from-build-artifacts) to test the artifacts.

**Comment by JoostK** [comment]
I work on Angular in my spare time, so that includes weekends ;-) But no hurry from my side, you're fine to follow up whenever fits you (please do take weekends off, haha).

**Comment by JoostK** [comment]
Here's the artifacts: https://app.circleci.com/pipelines/github/angular/angular/42604/workflows/4a98e0a3-1449-4f04-ae90-437222583045/jobs/1121108/artifacts

You'll only need `@angular/animations` as the changes are exclusive to that package.

**Comment by JonWallsten** [comment]
Haha, I'll test tomorrow then! I'll comment here if I have trouble installing the artifacts or else I report back the result! Thanks!

**Comment by JonWallsten** [comment]
@JonWallsten: It seems like it went from 1.5 seconds to 5 milliseconds (!) How to I verify that everything is correct? I'm not sure what the code does?
![image](https://user-images.githubusercontent.com/5993909/153701415-41122c9e-0282-4ce1-8ae6-2d12542f5a49.png)

**Comment by JonWallsten** [comment]
Thank you for the explanation! You're are correct that I still not fully understand, but it still makes sense.
It sounds like you guys can verify this without needing to have my exact code or data, so that's good.
Any idea why the traverse was like this? And why my use case triggered it? Is it because we have a pretty deep hierarchy of components and literally thousands of material components?

**Comment by JoostK** [comment]
Not sure why it is like this; I suspect it's just from the perspective of needing to find an insertion point that you iterate over all potential insertion points to find the correct index, not realizing the algorithmic complexity of this approach.

In the meantime I've looked at an application where I can reproduce the issue, it simply occurs when there's lots of animated components on a page which is easily achieved using just `ngFor`.

**Comment by JonWallsten** [comment]
Ah, I guess the implementer didn't take us into account. We have a lot of nested *ngFor loops when we render sections and tables consisting of components with menus, autocomplete, form controls and stuff. Many rows and sometimes many columns. It adds up in the end!

**Comment by Miwa199739** [comment]
@JoostK, we hit the same issue with "@angular/animations": "13.1.3", and thanks for fixing this! 
Could you please let us know when will your commit be rolled out as a stable @angular/animations package? 
Our PROD deployment is actually blocked by this regression, and could you please confirm if it's safe to utilize this artifact https://app.circleci.com/pipelines/github/angular/angular/42604/workflows/4a98e0a3-1449-4f04-ae90-437222583045/jobs/1121108/artifacts to short-term unblock PROD deployment? And we'll switch to the newly-released animation package once your change is rolled out.

**Comment by Miwa199739** [comment]
@JoostK, updated animations package to reference the artifact you provided, and npm install -> npm start, but we hit the following issue:
![image](https://user-images.githubusercontent.com/14820833/154623980-84956e4e-2ff4-407a-8410-3f9e584df8df.png)

package.json:

  "dependencies": {
    "@angular-react/core": "4.0.2",
    "@angular-react/fabric": "4.0.2",
    "@angular/animations": "https://1121108-24195339-gh.circle-artifacts.com/0/angular/animations-pr45057-d4fd692abf.tgz",
    "@angular/cdk": "^13.1.3",
    "@angular/common": "13.1.3",
    "@angular/compiler": "13.1.3",
    "@angular/core": "13.1.3",
    "@angular/forms": "13.1.3",
    "@angular/material": "^13.1.3",
    "@angular/platform-browser": "13.1.3",
    "@angular/platform-browser-dynamic": "13.1.3",
    "@angular/router": "13.1.3"
}

**Comment by JoostK** [comment]
The artifacts from #45057 are based on v14-next, so those are unlikely to be compatible with 13.1.3. The artifacts from #45113 are based on a commit close to the 13.2.3 release commit, so if you intent to use snapshot artifacts then you'd have to at least use Angular 13.2.3, or use _all_ `@angular` artifacts from that PR to ensure everything is in sync. It goes without saying that using snapshot artifacts is at your own risk.

**Comment by Miwa199739** [comment]
hi @JoostK, thanks for the response! 
We ended up with temporarily patching angular animation package with the fix in your commit: https://github.com/angular/angular/pull/45057, and will lift up the angular version once it gets released, thanks!

**Comment by JoostK** [comment]
This is now available on NPM in 13.2.4.
